STATE-COMPATIBILITY.md

```diff
@@ -0,0 +1,211 @@
+# State-Compatibility
+
+- [State-Compatibility](#state-compatibility)
+  - [Scope](#scope)
+  - [Validating State-Compatibility](#validating-state-compatibility)
+    - [AppHash](#apphash)
+    - [LastResultsHash](#lastresultshash)
+  - [Major Sources of State-incompatibility](#major-sources-of-state-incompatibility)
+    - [Creating Additional State](#creating-additional-state)
+    - [Changing Proto Field Definitions](#changing-proto-field-definitions)
+    - [Returning Different Errors Given Same Input](#returning-different-errors-given-same-input)
+    - [Variability in Gas Usage](#variability-in-gas-usage)
+  - [Secondary Limitations To Keep In Mind](#secondary-limitations-to-keep-in-mind)
+    - [Network Requests to External Services](#network-requests-to-external-services)
+    - [Randomness](#randomness)
+    - [Parallelism and Shared State](#parallelism-and-shared-state)
+    - [Hardware Errors](#hardware-errors)
+
+
+It is critical for the patch and minor releases to be state-machine compatible with prior releases in the same minor version. 
+For example, v13.0.2 must be state-machine compatible with v13.0.1. 
+_An exception are minor releases that are either emergency releases or replacements of deprecated major releases_. 
+
+This is to ensure **determinism**, i.e., given the same input, the nodes will always produce the same output.
+
+State-incompatibility is allowed for major upgrades because all nodes in the network perform it at the same time. Therefore, after the upgrade, the nodes continue functioning in a deterministic way.
+ 
+## Scope
+
+The state-machine scope includes the following areas:
+
+- All ICS messages including:
+  - Every msg's ValidateBasic method
+  - Every msg's MsgServer method
+    - Net gas usage, in all execution paths
+    - Error result returned
+    - State changes (namely every store write)
+- AnteHandlers in "DeliverTx" mode
+- All `BeginBlock`/`EndBlock` logic
+
+The following are **NOT** in the state-machine scope:
+
+- Events
+- Queries that are not whitelisted
+- CLI interfaces
+
+## Validating State-Compatibility 
+
+CometBFT ensures state compatibility by validating a number of hashes that can be found [here](https://github.com/cometbft/cometbft/blob/9f76e8da150414ce73eed2c4f248947b657c7587/proto/tendermint/types/types.proto#L70-L77).
+
+`AppHash` and `LastResultsHash` are the common sources of problems stemming from our work.
+To avoid these problems, let's now examine how these hashes work.
+
+### AppHash
+
+**Note:** The following explanation is simplified for clarity.
+
+An app hash is a hash of hashes of every store's Merkle root that is returned by ABCI's `Commit()` from Cosmos-SDK to CometBFT.
+Cosmos-SDK [takes an app hash of the application state](https://github.com/osmosis-labs/cosmos-sdk/blob/5c9a51c277d067e0ec5cf48df30a85fae95bcd14/store/rootmulti/store.go#L430), and propagates it to CometBFT which, in turn, compares it to the app hash of the rest of the network.
```

Comment:
Any reason why the link points to the Osmosis' fork here?

Reply:
Fixed it https://github.com/cosmos/gaia/pull/2818/commits/d96465f02e07dc17ad65e2c5e6fe0fd057d7bd83